File: app/src/main/java/com/greenbox/coyni/view/CustomKeyboard.java
package com.greenbox.coyni.view;

import android.content.Context;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.util.SparseArray;
import android.view.LayoutInflater;
import android.view.View;
import android.view.inputmethod.InputConnection;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.greenbox.coyni.R;

public class CustomKeyboard extends LinearLayout implements View.OnClickListener {

    private TextView keyOne,keyTwo,keyThree,keyFour,keyFive,keySix,keySeven,keyEight,keyNine,keyZero,keyDot,keyActionText;
    private LinearLayout keyBack,keyAction;
    private SparseArray<String> keyValues=new SparseArray<>();
    InputConnection inputConnection;
    public CustomKeyboard(Context context){
        this(context,null,0);

    }

    public CustomKeyboard(Context context, AttributeSet attrs){
        this(context,attrs,0);

    }

    public CustomKeyboard(Context context, AttributeSet attrs, int defStyleattr){
        super(context,attrs,defStyleattr);
        init(context,attrs);

    }

    private void init(Context context,AttributeSet attributeSet){
        LayoutInflater.from(context).inflate(R.layout.activity_custom_keyboard,this,true);
        keyOne=findViewById(R.id.keyOneTV);
        keyOne.setOnClickListener(this);

        keyTwo=findViewById(R.id.keyTwoTV);
        keyTwo.setOnClickListener(this);

        keyThree=findViewById(R.id.keyThreeTV);
        keyThree.setOnClickListener(this);

        keyFour=findViewById(R.id.keyFourTV);
        keyFour.setOnClickListener(this);

        keyFive=findViewById(R.id.keyFiveTV);
        keyFive.setOnClickListener(this);

        keySix=findViewById(R.id.keySixTV);
        keySix.setOnClickListener(this);

        keySeven=findViewById(R.id.keySevenTV);
        keySeven.setOnClickListener(this);

        keyEight=findViewById(R.id.keyEightTV);
        keyEight.setOnClickListener(this);

        keyNine=findViewById(R.id.keyNineTV);
        keyNine.setOnClickListener(this);

        keyZero=findViewById(R.id.keyZeroTV);
        keyZero.setOnClickListener(this);

        keyDot=findViewById(R.id.keyDotTV);
        keyDot.setOnClickListener(this);

        keyBack=findViewById(R.id.keyBackLL);
        keyBack.setOnClickListener(this);

        keyAction=findViewById(R.id.keyActionLL);
        keyAction.setOnClickListener(this);

        keyActionText=findViewById(R.id.keyActionTV);

        keyValues.put(R.id.keyZeroTV,"0");
        keyValues.put(R.id.keyOneTV,"1");
        keyValues.put(R.id.keyTwoTV,"2");
        keyValues.put(R.id.keyThreeTV,"3");
        keyValues.put(R.id.keyFourTV,"4");
        keyValues.put(R.id.keyFiveTV,"5");
        keyValues.put(R.id.keySixTV,"6");
        keyValues.put(R.id.keySevenTV,"7");
        keyValues.put(R.id.keyEightTV,"8");
        keyValues.put(R.id.keyNineTV,"9");
        keyValues.put(R.id.keyDotTV,".");
        keyValues.put(R.id.keyActionLL,"");


    }
    @Override
    public void onClick(View view) {
        if (inputConnection == null) {
            CharSequence selectedText = inputConnection.getSelectedText(0);
            if (TextUtils.isEmpty(selectedText)) {
                inputConnection.deleteSurroundingText(1, 0);

            } else {
                inputConnection.commitText("", 1);
            }
        } else {
            String value = keyValues.get(view.getId());
            try{
                inputConnection.commitText(value, 1);
            }catch (Exception e){
//                e.printStackTrace();
            }
        }

        keyBack.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View view) {
                String chatSet = (String) inputConnection.getSelectedText(0);
                try {
                    inputConnection.deleteSurroundingText(1,0);
                }catch (Exception e){
//                    e.printStackTrace();
                }
            }
        });


    }

    public void setInputConnection(InputConnection ic){
        inputConnection=ic;
    }

    public void setKeyAction(String actionName){
        keyActionText.setText(actionName);
    }

}

File: app/src/main/java/com/coyni/pos/app/utils/keyboards/CustomKeyboard.kt
package com.coyni.pos.app.utils.keyboards

import android.content.Context
import android.os.Build
import android.text.TextUtils
import android.util.AttributeSet
import android.util.SparseArray
import android.view.LayoutInflater
import android.view.View
import android.view.inputmethod.InputConnection
import android.widget.LinearLayout
import androidx.annotation.RequiresApi
import com.coyni.pos.app.R
import com.coyni.pos.app.databinding.CustomKeyboardBinding
import com.coyni.pos.app.utils.Utils
import com.coyni.pos.app.view.RefundTransactionActivity
import java.lang.Exception

class CustomKeyboard : LinearLayout, View.OnClickListener {

    private val keyValues = SparseArray<String>()
    lateinit var listener: OnSuccessListener
    lateinit var keybaackListner: OnKeyBackSuccessListener
    var enteredText: String = ""
    lateinit var binding: CustomKeyboardBinding
    private lateinit var inputConnection: InputConnection
    lateinit var screen: String

    @RequiresApi(Build.VERSION_CODES.M)
    constructor(context: Context) : super(context) {
//        context = context
        initView(context)
    }

    @RequiresApi(Build.VERSION_CODES.M)
    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        initView(context)
    }

    interface OnSuccessListener {
        fun onKeyboardClick(action: String, value: String?);
    }

    interface OnKeyBackSuccessListener {
        fun onKeybackClick(action: String, value: String?);
    }

    fun setOnKeyboardClickListener(listener: OnSuccessListener) {
        this.listener = listener
    }

    fun setOnKeybackClickListener(listener: OnKeyBackSuccessListener) {
        this.keybaackListner = listener
    }

    private fun initView(context: Context) {
//        inflate(getContext(), R.layout.custom_keyboard, this)
        val inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        binding = CustomKeyboardBinding.inflate(inflater, this, true)
        clickListners()
        setupKeyvalues();
    }

    private fun clickListners() {
        binding.keyOneTV.setOnClickListener(this)
        binding.keyTwoTV.setOnClickListener(this)
        binding.keyThreeTV.setOnClickListener(this)
        binding.keyFourTV.setOnClickListener(this)
        binding.keyFiveTV.setOnClickListener(this)
        binding.keySixTV.setOnClickListener(this)
        binding.keySevenTV.setOnClickListener(this)
        binding.keyEightTV.setOnClickListener(this)
        binding.keyNineTV.setOnClickListener(this)
        binding.keyZeroTV.setOnClickListener(this)
        binding.dotTV.setOnClickListener(this)

        binding.keyActionLL.setOnClickListener(OnClickListener {
            if (listener != null) {
                listener.onKeyboardClick(Utils.BUTTON_CLICK, "")
            }
        })
        binding.keyBackLL.setOnClickListener(View.OnClickListener {
            val charSet = inputConnection.getSelectedText(0)
            try {
                inputConnection.deleteSurroundingText(1, 0)
                if (enteredText.length > 1) {
                    enteredText = enteredText.substring(0, enteredText.length - 1)
                }
                if (keybaackListner != null) {
                    keybaackListner.onKeybackClick(Utils.REFUND, "")
                }

            } catch (e: Exception) {
                e.printStackTrace()
            }
        })
    }

    private fun setupKeyvalues() {
        keyValues.put(R.id.keyZeroTV, "0")
        keyValues.put(R.id.keyOneTV, "1")
        keyValues.put(R.id.keyTwoTV, "2")
        keyValues.put(R.id.keyThreeTV, "3")
        keyValues.put(R.id.keyFourTV, "4")
        keyValues.put(R.id.keyFiveTV, "5")
        keyValues.put(R.id.keySixTV, "6")
        keyValues.put(R.id.keySevenTV, "7")
        keyValues.put(R.id.keyEightTV, "8")
        keyValues.put(R.id.keyNineTV, "9")
        keyValues.put(R.id.dotTV, ".")
        keyValues.put(R.id.keyActionCV, "")
    }

    override fun onClick(v: View?) {
        if (inputConnection == null) {
            val selectedText: CharSequence = inputConnection.getSelectedText(0)
            if (TextUtils.isEmpty(selectedText)) {
                inputConnection.deleteSurroundingText(1, 0)
            } else {
                inputConnection.commitText("", 1)
            }
        }
        if (enteredText.length < 8) {
            val value: String = keyValues.get(v!!.id)
            if ((enteredText == "" || enteredText.contains(".")) && value == ".") {
            } else {
                println(enteredText)
                enteredText = enteredText + value
                inputConnection.commitText(value, 1)
                if (keybaackListner != null) {
                    keybaackListner.onKeybackClick(Utils.REFUND, "")
                }
            }
        }
    }

    fun setInputConnection(ic: InputConnection) {
        inputConnection = ic
    }

    fun setKeyAction(actionName: String?, context: Context?) {
        binding.keyActionTV.text = actionName;
    }


    fun setScreenName(screenName: String) {
//        screen = screenName
    }

    fun enableButton() {
        binding.keyActionLL.setBackgroundResource(R.drawable.custom_keyboard_action_btn_enable_bg)
        binding.keyActionLL.setEnabled(true)
    }

    fun disableButton() {
        binding.keyActionLL.setBackgroundResource(R.drawable.custom_keyboard_action_btn_disable_bg)
        binding.keyActionLL.setEnabled(false)
    }

    fun clearData() {
        enteredText = ""
    }

    fun setText(strText: String) {
        enteredText = strText
    }

    @JvmName("setEnteredText1")
    fun setEnteredText(text: String) {
        enteredText = text.trim { it <= ' ' }
    }
}

